notebooks/02_model/fm_deep_dive.ipynb

```diff
@@ -0,0 +1,894 @@
+       "  <thead>\n",
+       "    <tr style=\"text-align: right;\">\n",
+       "      <th></th>\n",
+       "      <th>rating</th>\n",
+       "      <th>field1</th>\n",
+       "      <th>field2</th>\n",
+       "      <th>field3</th>\n",
+       "      <th>field4</th>\n",
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "    <tr>\n",
+       "      <td>0</td>\n",
+       "      <td>1</td>\n",
+       "      <td>1:1:1</td>\n",
+       "      <td>2:4:3</td>\n",
+       "      <td>3:5:1.0</td>\n",
+       "      <td>4:6:1</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <td>1</td>\n",
+       "      <td>0</td>\n",
+       "      <td>1:2:1</td>\n",
+       "      <td>2:4:4</td>\n",
+       "      <td>3:5:2.0</td>\n",
+       "      <td>4:7:1</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <td>2</td>\n",
+       "      <td>0</td>\n",
+       "      <td>1:3:1</td>\n",
+       "      <td>2:4:5</td>\n",
+       "      <td>3:5:3.0</td>\n",
+       "      <td>4:8:1</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <td>3</td>\n",
+       "      <td>1</td>\n",
+       "      <td>1:3:1</td>\n",
+       "      <td>2:4:6</td>\n",
+       "      <td>3:5:4.0</td>\n",
+       "      <td>4:9:1</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <td>4</td>\n",
+       "      <td>1</td>\n",
+       "      <td>1:3:1</td>\n",
+       "      <td>2:4:7</td>\n",
+       "      <td>3:5:5.0</td>\n",
+       "      <td>4:10:1</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "   rating field1 field2   field3  field4\n",
+       "0       1  1:1:1  2:4:3  3:5:1.0   4:6:1\n",
+       "1       0  1:2:1  2:4:4  3:5:2.0   4:7:1\n",
+       "2       0  1:3:1  2:4:5  3:5:3.0   4:8:1\n",
+       "3       1  1:3:1  2:4:6  3:5:4.0   4:9:1\n",
+       "4       1  1:3:1  2:4:7  3:5:5.0  4:10:1"
+      ]
+     },
+     "execution_count": 2,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "df_feature_original = pd.DataFrame({\n",
+    "    'rating': [1, 0, 0, 1, 1],\n",
+    "    'field1': ['xxx1', 'xxx2', 'xxx4', 'xxx4', 'xxx4'],\n",
+    "    'field2': [3, 4, 5, 6, 7],\n",
+    "    'field3': [1.0, 2.0, 3.0, 4.0, 5.0],\n",
+    "    'field4': ['1', '2', '3', '4', '5']\n",
+    "})\n",
+    "\n",
+    "converter = LibffmConverter().fit(df_feature_original, col_rating='rating')\n",
+    "df_out = converter.transform(df_feature_original)\n",
+    "df_out"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are in total 4 fields and 10 features.\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('There are in total {0} fields and {1} features.'.format(converter.field_count, converter.feature_count))"
```

Comment:
In general, would be nice to have black-like formatting (using blackmagic maybe). But formatting is not super important so it is just a suggestion.

Reply:
added issue to work on it when we have time https://github.com/microsoft/recommenders/issues/1000